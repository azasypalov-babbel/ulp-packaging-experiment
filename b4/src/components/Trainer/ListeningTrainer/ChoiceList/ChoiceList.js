import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import ChoiceItem from '../../../shared/ChoiceItem/ChoiceItem';
import { PoseGroup } from 'react-pose';

/*
  This placeholder component reserves space in the stage before
  the items are rendered so that the vertical centering is calculated correctly.
*/
const StyledPlaceholderList = styled.div`
  display: flex;
  width: 100%;
  align-items: center;
  flex-direction: column;
  min-height: ${(props) => `${props.choiceItemCount * 3.75}rem`};

  & > * + * {
    margin-top: 0.75rem;
  }
`;

const ChoiceList = ({ items, showChoiceItems, onClick, attemptedText, correctText }) => (
  <StyledPlaceholderList
    data-selector="choicelist-container"
    data-has-attempt={Boolean(attemptedText)}
    choiceItemCount={items.length}
  >
    <PoseGroup flipMove={false} animateOnMount>
      {items.map((choiceItemText, index) => {
        const hasAttempt = Boolean(attemptedText);

        const isCorrectItem = choiceItemText === correctText;
        const dataSelector = isCorrectItem ? 'correct-item' : 'distractor';
        const isMistakenItem = choiceItemText === attemptedText && attemptedText !== correctText;

        return (
          showChoiceItems
            ? <ChoiceItem
              animationDelayOffset={index}
              key={`${correctText}_${choiceItemText}`}
              onClick={(evt) => onClick(choiceItemText, evt)}
              dataSelector={dataSelector}
              data-selector="display-language-text"
              listenToKey={`${index + 1}`}
              correctFeedback={hasAttempt && isCorrectItem}
              mistakeFeedback={hasAttempt && isMistakenItem}
              disabled={hasAttempt}
            >
              {choiceItemText}
            </ChoiceItem>
            : null
        );
      })}
    </PoseGroup>
  </StyledPlaceholderList>
);

ChoiceList.propTypes = {
  correctText: PropTypes.string.isRequired,
  items: PropTypes.array.isRequired,
  onClick: PropTypes.func.isRequired,
  attemptedText: PropTypes.string,
  showChoiceItems: PropTypes.bool.isRequired
};

export default ChoiceList;
