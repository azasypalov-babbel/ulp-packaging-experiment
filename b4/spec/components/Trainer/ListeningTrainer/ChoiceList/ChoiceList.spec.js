import React from 'react';
import ChoiceList from '../../../../../src/components/Trainer/ListeningTrainer/ChoiceList/ChoiceList';
import { shallow } from 'enzyme';

describe('<ChoiceList />', () => {
  const defaultProps = {
    showChoiceItems: true,
    correctText: 'One',
    items: ['Two', 'Three', 'Four', 'Five'],
    onClick: () => {}
  };

  test('renders', () => {
    const component = shallow(<ChoiceList {...defaultProps} />);
    expect(component).toMatchSnapshot();
  });

  describe('with correct attempt', () => {
    test('renders', () => {
      const component = shallow(<ChoiceList {...defaultProps} attemptedText="One" />);
      expect(component).toMatchSnapshot();
    });
  });

  describe('with incorrect attempt', () => {
    test('renders', () => {
      const component = shallow(<ChoiceList {...defaultProps} attemptedText="Two" />);
      expect(component).toMatchSnapshot();
    });
  });
});
