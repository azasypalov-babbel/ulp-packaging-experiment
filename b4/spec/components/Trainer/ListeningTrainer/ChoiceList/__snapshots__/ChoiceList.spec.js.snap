// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChoiceList /> renders 1`] = `
<styled.div
  choiceItemCount={4}
  data-has-attempt={false}
  data-selector="choicelist-container"
>
  <PoseGroup
    animateOnMount={true}
    flipMove={false}
  >
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={0}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={false}
      key="One_Two"
      listenToKey="1"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Two
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={1}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={false}
      key="One_Three"
      listenToKey="2"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Three
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={2}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={false}
      key="One_Four"
      listenToKey="3"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Four
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={3}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={false}
      key="One_Five"
      listenToKey="4"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Five
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
  </PoseGroup>
</styled.div>
`;

exports[`<ChoiceList /> with correct attempt renders 1`] = `
<styled.div
  choiceItemCount={4}
  data-has-attempt={true}
  data-selector="choicelist-container"
>
  <PoseGroup
    animateOnMount={true}
    flipMove={false}
  >
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={0}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={true}
      key="One_Two"
      listenToKey="1"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Two
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={1}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={true}
      key="One_Three"
      listenToKey="2"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Three
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={2}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={true}
      key="One_Four"
      listenToKey="3"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Four
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={3}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={true}
      key="One_Five"
      listenToKey="4"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Five
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
  </PoseGroup>
</styled.div>
`;

exports[`<ChoiceList /> with incorrect attempt renders 1`] = `
<styled.div
  choiceItemCount={4}
  data-has-attempt={true}
  data-selector="choicelist-container"
>
  <PoseGroup
    animateOnMount={true}
    flipMove={false}
  >
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={0}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={true}
      key="One_Two"
      listenToKey="1"
      mistakeFeedback={true}
      onClick={[Function]}
    >
      Two
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={1}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={true}
      key="One_Three"
      listenToKey="2"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Three
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={2}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={true}
      key="One_Four"
      listenToKey="3"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Four
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
    <WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))
      animationDelayOffset={3}
      correctFeedback={false}
      data-selector="display-language-text"
      dataSelector="distractor"
      disabled={true}
      key="One_Five"
      listenToKey="4"
      mistakeFeedback={false}
      onClick={[Function]}
    >
      Five
    </WithKeypress(Connect(WithShortcutHint(WithAnimations(ChoiceItem))))>
  </PoseGroup>
</styled.div>
`;
